### PR TITLE
Fix DSpark + DeepSeek V4 prefill CP compatibility

### PR DESCRIPTION
**Pull request by whybeyoung**
## Summary
Two related fixes so DSpark speculative decoding works together with DeepSeek V4 prefill context parallelism (`--enable-nsa-prefill-context-parallel --nsa-prefill-cp-mode round-robin-split`) at PD co-located deployment.

## Changes
1. **`deepseek_v4_dspark.py`: use lm_head's shard group for markov_w2 TP-shard**
   `configure_tp_shard` hard-coded `attn_tp_group` for the per-step all-gather. Under prefill CP (`attn_cp_size == tp_size`) `attn_tp_group` degenerates to `world_size == 1` while `lm_head` still shards over the full TP group, so the geometry check failed and the gather (if forced through) would be a no-op leaving half the vocab missing. Now we pick the group that matches `lm_head`'s own choice: `attn_tp_group` when `use_attn_tp_group=True`, otherwise the full TP group.

2. **`deepseek_v4.py`: enable DSpark aux capture under prefill CP**
   Drop the `NotImplementedError` blocking DSpark + CP and all-gather + rerange each `dspark_aux_hidden_states` tensor on the last PP rank with the same CP token split used for `hidden_states`, so downstream `logits_processor` indexing sees the global token layout.

## Test
Verified with:
```
--tp 8 --speculative-algorithm DSPARK
--enable-nsa-prefill-context-parallel --nsa-prefill-cp-mode round-robin-split
--page-size 256 ...
```
No more `NotImplementedError`; markov_w2 TP-shard geometry check passes.

---
### CI States

Latest PR Test (Base): :x: [Run #31455444987](https://github.com/sgl-project/sglang/actions/runs/31455444987)
Latest PR Test (Extra): :x: [Run #31455444860](https://github.com/sgl-project/sglang/actions/runs/31455444860)

CC  @xu-yfei @ShangmingCai @Fridge003

### CODE REVIEW

File: python/sglang/srt/models/deepseek_v4.py [hunk 1]
```diff
@@ -2445,12 +2439,21 @@ def forward(
...
             hidden_states = cp_all_gather_rerange_output(
                 hidden_states,
                 self.cp_size,
                 forward_batch,
-                torch.cuda.current_stream(),
+                stream,
             )
+            # Gather DSpark aux tensors on the same CP token split.
+            if capture_dspark:
+                dspark_aux_hidden_states = [
```
Comment (hzh0425) [inline python/sglang/srt/models/deepseek_v4.py:2554]: should we add an e2e test for dsv4 cp + dspark?

**Review by hzh0425 — APPROVED** [review]

**Comment by Fridge003** [comment]
@whybeyoung  We have shifted the CP implementation of DSV4 to CP_V2 in https://github.com/sgl-project/sglang/pull/33532. The older CP implementation will be deprecated soon.

Can you please test how it works with DSpark

Its usage should be:
```bash
--attn-cp-size 8 \
--enable-prefill-cp \
--cp-strategy interleave \
```
with flag SGLANG_ENABLE_CP_V2=1

File: test/registered/cp/test_deepseek_v4_dspark_cp.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
+"""E2E: DeepSeek-V4-Flash DSPARK speculation together with DSA prefill CP.
```
Comment (Fridge003) [inline test/registered/cp/test_deepseek_v4_dspark_cp.py]: No need to create a new test file. Can we add a subtest for test/registered/cp/test_deepseek_v4_flash_fp4_b200_cp.py

**Review by Fridge003 — DISMISSED** [review]

**Comment by whybeyoung** [comment]
Addressed review comments and rebased on latest main:

**@Fridge003 - CP_V2 support:**
- Extended `_execute_extend_cp_v2` in `eager_runner.py` to also gather `aux_hidden_states` via `cp_gather_after_forward`, mirroring the hidden-states gather; without it DSpark aux tensors stay CP-local while logits_processor indexes them with global `extend_seq_lens`.
- The in-model gather in `deepseek_v4.py` is now gated on `use_prefill_cp and not cp_v2_active` (matches upstream), so CP_V1 still works and CP_V2 uses the new external gather.

**@Fridge003 - test consolidation:**
- Removed the standalone `test_deepseek_v4_dspark_cp.py`.
- Added `TestDSV4FlashFP4B200_CP_DSpark` as a subtest in `test_deepseek_v4_flash_fp4_b200_cp.py` using the CP_V2 flag set (`--enable-prefill-cp --cp-strategy interleave`, `SGLANG_ENABLE_CP_V2=1`).

Rebased onto upstream/main; merge conflicts (with #33532) resolved.

**Comment by Fridge003** [comment]
/rerun-test test/registered/cp/test_gqa_prefill_cp.py  test/registered/cp/test_deepseek_v4_flash_fp4_b200_cp.py test/registered/cp/test_cp_strategy_unit.py  test/registered/cp/test_dsa_prefill_cp.py  test/registered/cp/test_deepseek_v3_cp_single_node.py

**Comment by Fridge003** [comment]
/rerun-test test/registered/dcp/test_kimi_linear_dcp_dspark4.py

**Review by Fridge003 — APPROVED** [review]

**Comment by whybeyoung** [comment]
/rerun-test test/registered/cp/test_gqa_prefill_cp.py test/registered/cp/test_deepseek_v4_flash_fp4_b200_cp.py test/registered/cp/test_cp_strategy_unit.py test/registered/cp/test_dsa_prefill_cp.py test/registered/cp/test_deepseek_v3_cp_single_node.py

**Comment by whybeyoung** [comment]
/rerun-test test/registered/dcp/test_kimi_linear_dcp_dspark4.py

**Comment by whybeyoung** [comment]
Pushed a follow-up fix (64ac5ae) for the b200 CP failure.

Root cause: in `_execute_extend_cp_v2` I was always forwarding `pre_hc_head` as `hidden_states_before_norm` to the logits processor. `DeepseekV4ForCausalLM.forward` only does that when aux capture is off; when DSpark aux capture is on, it passes `hidden_states_before_norm=None` so the logits processor stores the packed aux tensor (`[T, K*hidden]`) and hands it to the DSpark KV injector. Because `_get_hidden_states_to_store` prefers `hidden_states_before_norm` when present, my old CP-v2 code was silently swapping the packed aux for `pre_hc_head` (`[T, mult*hidden]`), which then tripped the `A.shape[-1] == B.shape[-1]` assertion inside `main_proj`'s fp8 matmul.

The CP-v2 code now mirrors the wrapper's contract: pass `hidden_states_before_norm` only when `aux_hidden_states is None`.

/rerun-test test/registered/cp/test_deepseek_v4_flash_fp4_b200_cp.py test/registered/dcp/test_kimi_linear_dcp_dspark4.py

**Comment by whybeyoung** [comment]
/rerun-test test/registered/cp/test_gqa_prefill_cp.py test/registered/cp/test_deepseek_v4_flash_fp4_b200_cp.py test/registered/cp/test_cp_strategy_unit.py test/registered/cp/test_dsa_prefill_cp.py test/registered/cp/test_deepseek_v3_cp_single_node.py

**Comment by whybeyoung** [comment]
/rerun-test test/registered/dcp/test_kimi_linear_dcp_dspark4.py

**Comment by whybeyoung** [comment]
/rerun-test test/registered/cp/test_gqa_prefill_cp.py test/registered/cp/test_deepseek_v4_flash_fp4_b200_cp.py test/registered/cp/test_cp_strategy_unit.py test/registered/cp/test_dsa_prefill_cp.py test/registered/cp/test_deepseek_v3_cp_single_node.py   test/registered/dcp/test_kimi_linear_dcp_dspark4.py
